pdfrw/objects/pdfstring.py

```diff
@@ -13,62 +77,83 @@ class PdfString(str):
-                     }
-    unescape_pattern = (r'(\\\\|\\b|\\f|\\n|\\r|\\t'
-                        r'|\\\r\n|\\\r|\\\n|\\[0-9]{3}|\\)')
-    unescape_func = re.compile(unescape_pattern).split
-
-    hex_pattern = '([a-fA-F0-9][a-fA-F0-9]|[a-fA-F0-9])'
-    hex_func = re.compile(hex_pattern).split
-
-    hex_pattern2 = ('([a-fA-F0-9][a-fA-F0-9][a-fA-F0-9][a-fA-F0-9]|'
-                    '[a-fA-F0-9][a-fA-F0-9]|[a-fA-F0-9])')
-    hex_func2 = re.compile(hex_pattern2).split
-
-    hex_funcs = hex_func, hex_func2
-
-    def decode_regular(self, remap=chr):
-        assert self[0] == '(' and self[-1] == ')'
-        mylist = self.unescape_func(self[1:-1])
-        result = []
-        unescape = self.unescape_dict.get
-        for chunk in mylist:
-            chunk = unescape(chunk, chunk)
-            if chunk.startswith('\\') and len(chunk) > 1:
-                value = int(chunk[1:], 8)
-                # FIXME: TODO: Handle unicode here
-                if value > 255:
-                    value = 255
-                chunk = remap(value)
-            if chunk:
-                result.append(chunk)
-        return ''.join(result)
-
-    def decode_hex(self, remap=chr, twobytes=False):
-        data = ''.join(self.split())
-        data = self.hex_funcs[twobytes](data)
-        chars = data[1::2]
-        other = data[0::2]
-        assert (other[0] == '<' and
-                other[-1] == '>' and
-                ''.join(other) == '<>'), self
-        return ''.join([remap(int(x, 16)) for x in chars])
-
-    def decode(self, remap=chr, twobytes=False):
-        if self.startswith('('):
-            return self.decode_regular(remap)
+    escapes = {
+        b'\n': b'\\n',
+        b'\r': b'\\r',
+        b'\t': b'\\t',
+        b'\b': b'\\b',
+        b'\f': b'\\f',
+        b'(' : b'\\(',
+        b')' : b'\\)',
+        b'\\': b'\\\\'
+    }
+
+    unescapes = {v[1]: k for k, v in escapes.items()}
+
+    # Possible string escapes from the spec:
+    # 1. \[nrtbf\()]: simple escapes
+    # 2. \\d{1,3}: octal. Must be zero-padded to 3 digits if followed by digit
+    # 3. \<end of line>: line continuation. We don't know the EOL marker
+    # used in the PDF, so accept \r, \n, and \r\n.
+    unescape_re = re.compile(br'\\(([nrtbf\(\)\\])|(\d{1,3})|(\r|\r\n|\n))')
+
+    def decode_escaped(self):
+        def handle_escape(m):
+            if m.group(2):  # direct escape
+                return self.unescapes[m.group(2).decode('ascii')]
+            if m.group(3):  # octal
+                value = int(m.group(3).decode('ascii'), 8)
+                return bytes(bytearray([value]))
+            if m.group(4):  # line continuation
+                return b''
+            # The PDF spec says that if the backslash doesn't form a valid
+            # escape, the backslash is ignored.
+            return m.group(1)
+
+        return self.unescape_re.sub(handle_escape, bytearray(self[1:-1]))
```

Comment:
I agree that `PdfString` would ideally inherit from `bytes`. I did try going down that path and as I recall ran into problems with breakages elsewhere. I think it might be worth working through them precisely to ensure that no consumer is bypassing the encoding/decoding accidentally.

@pmaupin, can you explain a little more what you mean by leakage here?